File: 28-implement-strstr/28-implement-strstr.cpp
// O(n) solution \U0001f525
class Solution {
public:
    vector<int> z;
    
    void populateZArray(string s){
        // initialise l, r and n
        int l = 0,r=0,n = s.size();
        
        // initialise z vector
        z = vector<int>(n);
        
        // from index 1 to n-1 , find length of prefix match found at position i
        for(int i=1;i<s.size();i++){
            
            // take boost if you are in the l to r zone
            if(i<=r){
                z[i] = min(r-i+1,z[i-l]);
            }
            // increment z[i] if current matches happen
            while(i+z[i]<n&&s[z[i]]==s[i+z[i]]){
                z[i]++;
            }
            
            // reconfigure l and r if you have exceeded current boundary
            if(i+z[i]-1>r){
                r = i+z[i]-1;
                l = i;
            }
            
        }
        
    }
    
    int strStr(string haystack, string needle) {
        if(needle=="") return 0;
        
        // add a delimiter between needle and haystack
        // this delimiter should not be part of character space of 
        // given input
        string temp = needle+"#"+haystack;
        
        // call z-function
        populateZArray(temp);
        
        // inside haystack find those places whose z value 
        // is equal to length of needle
        for(int i=needle.size()+1;i<temp.size();++i){
            if(z[i]==needle.size()){
                return i-needle.size()-1;    
            }
        }
        
        // if no match return -1
        return -1;
    }
};

File: 28-implement-strstr/28-implement-strstr.py
# KMP
# class Solution:
#     def strStr(self, haystack: str, needle: str) -> int:
#         def getLPS(s):
#             n = len(s)
#             lps = [0]*n
#             i,Len = 1,0
            
#             while i<n:
#                 if s[i]==s[Len]:
#                     Len+=1
#                     lps[i]=Len
#                     i+=1
#                 else:
#                     if Len>0:
#                         Len = lps[Len-1]
#                     else:
#                         lps[i]=0
#                         i+=1
#             return lps
#         kmp = getLPS(needle+"#"+haystack)
        
#         for i in range(len(kmp)):
#             if kmp[i]==len(needle):
#                 return i-len(needle)*2
#         return -1

class Solution:
    def strStr(self, haystack: str, needle: str) -> int:
        def getZArr(s):
            n = len(s)
            z = [0]*n
            l,r=0,0
            for i in range(1,n):
#                 take headstart from known region
                if i<=r:
                    z[i] = min(r-i+1,z[i-l])
#               get value from unknown region
                while i+z[i]<n and s[i+z[i]]==s[z[i]]:
                    z[i]+=1
            
#             expand known region
                if i+z[i]-1>r:
                    l = i
                    r = i+z[i]-1
            return z
        temp = needle+"#"+haystack
        zArr = getZArr(temp)
        print(temp)
        print(zArr)
        
        for idx in range(len(zArr)):
            if zArr[idx]==len(needle):
                return idx-len(needle)-1
        return -1